cmd/kaspawallet/keys/keys.go

```diff
@@ -0,0 +1,222 @@
+package keys
+
+import (
+	"bufio"
+	"crypto/cipher"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"github.com/kaspanet/kaspad/util"
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/argon2"
+	"golang.org/x/crypto/chacha20poly1305"
+	"os"
+	"path/filepath"
+	"runtime"
+)
+
+var (
+	defaultAppDir   = util.AppDir("kaspawallet", false)
+	defaultKeysFile = filepath.Join(defaultAppDir, "keys.json")
+)
+
+type keysFileJSON struct {
+	EncryptedPrivateKeys []string `json:"encryptedPrivateKeys"`
+	PublicKeys           []string `json:"publicKeys"`
+	MinimumSignatures    uint32   `json:"minimumSignatures"`
+}
+
+// Data holds all the data related to the wallet keys
+type Data struct {
+	encryptedPrivateKeys [][]byte
+	PublicKeys           [][]byte
+	MinimumSignatures    uint32
+}
+
+func (d *Data) toJSON() *keysFileJSON {
+	encryptedPrivateKeysHex := make([]string, len(d.encryptedPrivateKeys))
+	for i, encryptedPrivateKey := range d.encryptedPrivateKeys {
+		encryptedPrivateKeysHex[i] = hex.EncodeToString(encryptedPrivateKey)
+	}
+
+	publicKeysHex := make([]string, len(d.PublicKeys))
+	for i, publicKey := range d.PublicKeys {
+		publicKeysHex[i] = hex.EncodeToString(publicKey)
+	}
+
+	return &keysFileJSON{
+		EncryptedPrivateKeys: encryptedPrivateKeysHex,
+		PublicKeys:           publicKeysHex,
+		MinimumSignatures:    d.MinimumSignatures,
+	}
+}
+
+func (d *Data) fromJSON(kfj *keysFileJSON) error {
+	d.MinimumSignatures = kfj.MinimumSignatures
+
+	d.encryptedPrivateKeys = make([][]byte, len(kfj.EncryptedPrivateKeys))
+	for i, encryptedPrivateKey := range kfj.EncryptedPrivateKeys {
+		var err error
+		d.encryptedPrivateKeys[i], err = hex.DecodeString(encryptedPrivateKey)
+		if err != nil {
+			return err
+		}
+	}
+
+	d.PublicKeys = make([][]byte, len(kfj.PublicKeys))
+	for i, publicKey := range kfj.PublicKeys {
+		var err error
+		d.PublicKeys[i], err = hex.DecodeString(publicKey)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// DecryptPrivateKeys asks the user to enter the password for the private keys and
+// returns the decrypted private keys.
+func (d *Data) DecryptPrivateKeys() ([][]byte, error) {
+	password := getPassword("Password:")
+	aead, err := getAEAD(password)
+	if err != nil {
+		return nil, err
+	}
+
+	privateKeys := make([][]byte, len(d.encryptedPrivateKeys))
+	for i, encryptedPrivateKey := range d.encryptedPrivateKeys {
+		var err error
+		privateKeys[i], err = decryptPrivateKey(encryptedPrivateKey, aead)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return privateKeys, nil
+}
+
+// ReadKeysFile returns the data related to the keys file
+func ReadKeysFile(path string) (*Data, error) {
+	if path == "" {
+		path = defaultKeysFile
+	}
+
+	file, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+
+	decoder := json.NewDecoder(file)
+	decoder.DisallowUnknownFields()
+	decodedFile := &keysFileJSON{}
```

Comment:
Maybe also make `MinimumSignatures` a string? I'm a little scared about JSON numbers

Reply:
or if you try to encode a big integer, but that's probably only a risk with uint64